packages/nimble-components/src/table/specs/table-columns-hld.md

```diff
@@ -0,0 +1,287 @@
+    public slottedColumns: ITableColumn[] = [];
+
+    private slottedColumnsChanged(): void {
+        if (this.slottedColumns.length > 0) {
+            for (const tableColumn of this.slottedColumns) {
+                if (this.isTableColumn(tableColumn)) {
+                    // do init work
+                }
+            }
+        }
+    }
+}
+
+template:
+<template>
+    <div>
+        ...
+    </div>
+    <slot ${slotted('slottedColumns')}></slot>
+</template>
+```
+
+### `TableCellData` interface
+
+A table cell represents a single column for a single row. The data that a cell has access to will be a subset of the data for the entire row. An instance of a table cell will be generic to describe the subset of data it contains, where `TCellData` is a subset of the data represented by `TableRowData`:
+
+```TS
+interface TableCellData<TCellData extends TableRowData> {
+  data: TCellData;
+}
+```
+
+This interface could possibly be expanded in the future to communicate relevant table state to the cell template, such as whether or not the row is selected.
+
+### `ITableColumn` interface
+
+This interface is what a column web component (i.e. a slotted column) must implement.
+
+```TS
+interface ITableColumn<TCellData extends TableRowData> {
+    // The string to use in the header for the column
+    header?: string;
+
+    // The template to use for the header (string if none provided)
+    headerTemplateFn?: () => ViewTemplate;
+
+    // The template to use to render the cell content for the column
+    cellTemplate: ViewTemplate<TableCellData<TCellData>>;
+
+    // The keys that should be present in TCellData.
+    // This array is parallel with the keys returned from `getDataKeys()`.
+    readonly cellDataKeyNames: readonly string[];
+
+    // The keys from the row data that correlate to the data that will be in TCellData.
+    // This array is parallel with the keys specified by `dataKeyNames`.
+    getRowDataKeys: () => string[];
+
+    // Function that allows the table column to validate the type that gets created
+    // for the cell data. This should validate that the types in TCellData are correct
+    // for each key defined by `dataKeyNames`.
+    // Return `true` if the data is valid. Return `false` if the data is not valid.
+    validateCellData(cellData: TableRowData): boolean;
+}
+```
+
+The `ITableColumn` interface would be updated to support other features not covered in this HLD such as sorting and grouping.
+
+Given the above interface, a series of column providers to handle basic use cases can be written within nimble. For example, the `ITableColumn` implementation we could create for rendering data as a read-only `NimbleTextField` could look like this:
+
+```TS
+type TableColumnTextCellData = StringColumnData<'value'>;
+public TableColumnText extends FoundationElement implements ITableColumn<TableColumnTextCellData> {
+    ...
+
+    public cellDataKeyNames = ['value'] as const;
+
+    @attr
+    public valueKey: string;
```

Comment:
How is `valueKey` (renamed elsewhere to `data-key`?) an attribute if `cellDataKeyNames` hardcodes the key as "value"?

Reply:
This is the key a user provides to the column instance in the DOM (as an attribiute) to access the appropriate data out of the table data. The `cellDataKeyNames` are an implementation detail of an ITableColumn (something a user of the column would not be aware of), and are representative of the keyed values that are in the TableCellData passed to its `cellTemplate`.